cpp/src/search/multisearch.cu

```diff
@@ -0,0 +1,604 @@
+            index_range{range.end_, range.end_};
+    }
+
+    bool first_is_valid = cudf::util::bit_is_set(validity_pseudocolumn, range.start);
+    bool last_is_valid = cudf::util::bit_is_set(validity_pseudocolumn, range.last());
+
+    if (first_is_valid and last_is_valid) {
+        return index_range::empty_at(
+            nulls_appear_before_values ? range.start : range.end_
+        );
+    }
+
+    // At this point we know there's a transition from valid to invalid or vice-versa
+    // somewhere in the column, and we need to locate that point.
+
+    assert(logical_nxor(first_is_valid, nulls_appear_before_values));
+
+    index_range transition_range { range };
+        // Invariant: The first element of this range is before the transition, and the last element
+        // is after it. Which also means transition_range has at least 2 elements.
+
+    // Contract the transition search range using bisections
+
+    while (transition_range.length() > 2) {
+        auto middle_index = transition_range.middle();
+        auto middle_element_is_valid = cudf::util::bit_is_set(validity_pseudocolumn, middle_index);
+        auto middle_element_is_before_transition =
+            logical_nxor(nulls_appear_before_values, middle_element_is_valid);
+        transition_range = middle_element_is_before_transition ?
+            lower_half_and_middle(transition_range) :
+            upper_half(transition_range);
+    }
+    assert(transition_range.length() == 2);
+    return nulls_appear_before_values ?
+        index_range{ range.start, transition_range.last()} :
+        index_range{ transition_range.last(), range.end_};
+}
+
+
+/**
+ * @brief Try to continue narrowing the range of possible locations of the haystack
+ * element we're looking for (the first one greater than / greater-or-equal than
+ * the needle) using a single column.
+ *
+ * @note when this is run, we must have noticed that haystack records
+ * before @p equality_range.start are less-or-equal than / less than the
+ * needle - but we were using previous columns (in the haystack and needle) for that;
+ * similarly, records at @p equality_range.end_ - 1 and later are greater than
+ * the needle. It must be the case that the rest of the records in the range are
+ * identical to the needle up until this column.
+ *
+ * @note We assume the nullability matches, i.e. if the needle is null, then
+ * the column is nullable.
+ *
+ *
+ * @param[in] equality_range
+ *     the range of column indices whose elements we want to check for equality
+ *     with the needle element. Guaranteed to have more than a single element
+ *     when the function is called!
+ *
+ * @return The subrange of @p equality_range containing all elements which
+ * are order-equivalent to the needle, and if FindFirstGreater is true, also the
+ * first element coming after the needle, if one exists.
+ * po * first
+ */
+template <typename E>
+__device__ index_range  equality_subrange(
```

Comment:
Why can't we use `thrust::equal_range` and other Thrust binary search tools? Perhaps in combination with @kovaltan 's new iterators with null support? (#1833 )